arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/Either.kt

```diff
@@ -797,16 +797,16 @@ public sealed class Either<out A, out B> {
 
   public fun isLeft(): Boolean {
     contract {
-      returns(true) implies (this@Either is Left<A>)
-      returns(false) implies (this@Either is Right<B>)
+      returns(true) implies (this@Either is Left<*>)
+      returns(false) implies (this@Either is Right<*>)
```

Comment:
Why `*` instead of the type arguments? 🤔

Reply:
This is unrelated, indeed, but it turns out that the Kotlin compiler doesn't work well with contract with type arguments. If your types are named *exactly* like the ones in the contract, then everything is fine. But in the code the result has types `Either<R, A>` and `Either<R, B>`, and the compiler was becoming terribly confused. So I "fixed" the contracts in `Either`.

Reply:
It seems to have broken the smart-cast in `onLeft`, so does that not work anymore? That seems crucial functionality, should we report a bug? 🤔

Reply:
This PR can be implemented without that change, so I've reverted it